### PR TITLE
only use @bs.deriving, because of collisions with `ppxlib`

### PR DESCRIPTION
**Pull request by jfrolich**
Given the compatibility with OCaml it is probably also good to just keep `bs.deriving`.

### CODE REVIEW

**Comment by EduardoRFS** [comment]
There is some tests related to this, but I wonder if we could just ignore all the deriving that we don't know. So that we keep the compatibility, with current implementations.

What do you think @anmonteiro ? Is it worth to support only `@deriving`?

**Comment by anmonteiro** [comment]
There are / have been 2 issues related to `deriving`:

###  1. `[@deriving SOMETHING]` coming from (shared) native code

This first one was caused by the migration from `bs.attribute` to simply `attribute` done in https://github.com/rescript-lang/rescript-compiler/pull/4701. It started producing a hard failure when native code had a `@deriving` attribute, e.g. `@deriving yojson` in my [case](https://github.com/anmonteiro/gh-feed-reader/blob/80d2151621d0f9834d62033dfeb95e147549e797/shared/feed.ml)
  - @elliottcable fixed one of those hard failures in https://github.com/rescript-lang/rescript-compiler/pull/4911
  - I fixed the rest in [this commit](https://github.com/melange-re/melange/commit/3cd5a4df1ee95821aced316838aabc5e36c760c5) (towards the bottom of the diff)
  - I think what's still remaining to do for this case is to verify that `[@deriving yojson]` can be used in Melange with ppx_deriving_yojson, for example, and fix it if it doesn't work


### 2. `[@deriving abstract]` breaking PPXes compiled with ppxlib

This is the case @jfrolich identified. I'd love to get a deeper understanding of the issue here, and check whether it's possible to not raise a hard error. This is an action item for me: I'll play with some ppxes and ppxlib to gather more information about this.

The solution proposed in this PR effectively breaks forwards compatibility with the ReScript ecosystem. I knew this is going to be the case at _some_ point, which means we won't be a strict superset of ReScript, but I'm wondering if we can delay that as much as possible.

**Comment by ELLIOTTCABLE** [comment]
For whatever it's worth, I think it's best to break forward compatibility *now*. Or rather, put the onus of writing cross-compatible libraries on the library-author (like me! 👋) as early as possible.

Deferring pain like that sounds compassionate, but it's really not — all that time in the interim, you're allowing …

1. … new people to join the "going to have to deal with this, ugh" group (i.e. writing *new* cross-compatible libraries, but doing it in an old/incompatible way that will eventually break), and …
2. … both old and new library-authors to make design-choices which, were they better informed about the future path, they might not have made.

All that said, tooling or options that *improve* our ability to write cross-compatible code (even if it's finagley and diffictul and messy to do!) are very welcome. i.e. exactly that "ignore `@deriving` declarations we don't understand" plan.

tl;dr make good decisions now not better decisions later, even if it annoys some of us. we'll thank you for it. <3

**Comment by ELLIOTTCABLE** [comment]
(Above is exclusively taking about "forward" compatibility, though; let's please keep as much *backward* compatibility with existing BuckleScript-compatible code from before the fork as we possibly can. Please. Dear god. *Please.* 🤣)

**Comment by anmonteiro** [comment]
Thanks for the context, Elliott.

A few more thoughts on possible solutions:

- Have Melange register dummy PPX derivers with ppxlib for `jsConverter` and `abstract`
- run the Melange built-in transforms (which remove `deriving` attributes) before applying PPX transformations
  - I'm not immediately sure what the implications of this are, so it would need testing in a few codebases

**Comment by jfrolich** [comment]
I removed the backwards-incompatible code. Now it only prints with a namespace.

**Comment by anmonteiro** [comment]
I think it makes sense to keep printing `"bs.deriving"` -> `"bs.deriving"`. Let's move the remaining discussion over to #93, then.
@jfrolich it'd be great if you could try that one out in your codebase and report back.

**Comment by anmonteiro** [comment]
Aaaand, I fat-fingered the commit message. Oh well.
